// @remotion/studio/dist/components/RenderModal/SchemaEditor/ZodEnumEditor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZodEnumEditor = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const Checkmark_1 = require("../../../icons/Checkmark");
const ComboBox_1 = require("../../NewComposition/ComboBox");
const Fieldset_1 = require("./Fieldset");
const local_state_1 = require("./local-state");
const SchemaLabel_1 = require("./SchemaLabel");
const zod_schema_type_1 = require("./zod-schema-type");
const ZodFieldValidation_1 = require("./ZodFieldValidation");
const container = {
    width: '100%',
};
const ZodEnumEditor = ({ schema, jsonPath, setValue, defaultValue, value, onSave, showSaveButton, onRemove, saving, }) => {
    const { localValue, onChange: setLocalValue, reset, } = (0, local_state_1.useLocalState)({
        schema,
        setValue,
        unsavedValue: value,
        savedValue: defaultValue,
    });
    const enumValues = (0, zod_schema_type_1.getEnumValues)(schema);
    const isRoot = jsonPath.length === 0;
    const comboBoxValues = (0, react_1.useMemo)(() => {
        return enumValues.map((option) => {
            return {
                value: option,
                label: option,
                id: option,
                keyHint: null,
                leftItem: option === value ? jsx_runtime_1.jsx(Checkmark_1.Checkmark, {}) : null,
                onClick: (id) => {
                    setLocalValue(() => id, false, false);
                },
                quickSwitcherLabel: null,
                subMenu: null,
                type: 'item',
            };
        });
    }, [enumValues, setLocalValue, value]);
    const save = (0, react_1.useCallback)(() => {
        onSave(() => value, false, false);
    }, [onSave, value]);
    return (jsx_runtime_1.jsxs(Fieldset_1.Fieldset, { shouldPad: true, success: localValue.zodValidation.success, children: [
            jsx_runtime_1.jsx(SchemaLabel_1.SchemaLabel, { handleClick: null, onSave: save, showSaveButton: showSaveButton, isDefaultValue: localValue.value === defaultValue, onReset: reset, jsonPath: jsonPath, onRemove: onRemove, saving: saving, valid: localValue.zodValidation.success, saveDisabledByParent: !localValue.zodValidation.success, suffix: null }), jsx_runtime_1.jsx("div", { style: isRoot ? undefined : container, children: jsx_runtime_1.jsx(ComboBox_1.Combobox, { values: comboBoxValues, selectedId: value, title: value }) }), jsx_runtime_1.jsx(ZodFieldValidation_1.ZodFieldValidation, { path: jsonPath, localValue: localValue })
        ] }));
};
exports.ZodEnumEditor = ZodEnumEditor;
