// @remotion/studio/dist/components/RenderModal/SchemaEditor/ZodStaticFileEditor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ZodStaticFileEditor = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const Checkmark_1 = require("../../../icons/Checkmark");
const ComboBox_1 = require("../../NewComposition/ComboBox");
const use_static_files_1 = require("../../use-static-files");
const Fieldset_1 = require("./Fieldset");
const local_state_1 = require("./local-state");
const SchemaLabel_1 = require("./SchemaLabel");
const ZodFieldValidation_1 = require("./ZodFieldValidation");
const container = {
    width: '100%',
};
const ZodStaticFileEditor = ({ schema, jsonPath, setValue, defaultValue, value, onSave, showSaveButton, onRemove, saving, saveDisabledByParent, mayPad, }) => {
    const { localValue, onChange: setLocalValue, reset, } = (0, local_state_1.useLocalState)({
        schema,
        setValue,
        unsavedValue: value,
        savedValue: defaultValue,
    });
    const isRoot = jsonPath.length === 0;
    const staticFiles = (0, use_static_files_1.useStaticFiles)();
    const comboBoxValues = (0, react_1.useMemo)(() => {
        return staticFiles.map((option) => {
            return {
                value: option.src,
                label: option.name,
                id: option.src,
                keyHint: null,
                leftItem: option.src === value ? jsx_runtime_1.jsx(Checkmark_1.Checkmark, {}) : null,
                onClick: (id) => {
                    setLocalValue(() => id, false, false);
                },
                quickSwitcherLabel: null,
                subMenu: null,
                type: 'item',
            };
        });
    }, [setLocalValue, staticFiles, value]);
    const save = (0, react_1.useCallback)(() => {
        onSave(() => value);
    }, [onSave, value]);
    return (jsx_runtime_1.jsxs(Fieldset_1.Fieldset, { shouldPad: mayPad, success: localValue.zodValidation.success, children: [
            jsx_runtime_1.jsx(SchemaLabel_1.SchemaLabel, { handleClick: null, onSave: save, showSaveButton: showSaveButton, isDefaultValue: localValue.value === defaultValue, onReset: reset, jsonPath: jsonPath, onRemove: onRemove, saving: saving, valid: localValue.zodValidation.success, saveDisabledByParent: saveDisabledByParent, suffix: null }), jsx_runtime_1.jsx("div", { style: isRoot ? undefined : container, children: jsx_runtime_1.jsx(ComboBox_1.Combobox, { values: comboBoxValues, selectedId: localValue.value, title: value }) }), jsx_runtime_1.jsx(ZodFieldValidation_1.ZodFieldValidation, { path: jsonPath, localValue: localValue })
        ] }));
};
exports.ZodStaticFileEditor = ZodStaticFileEditor;
