// @remotion/studio-server/dist/preview-server/routes/can-update-sequence-props.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.computeSequencePropsStatusByLine = exports.computeSequencePropsStatus = exports.computeSequencePropsStatusFromContent = exports.lineColumnToNodePath = exports.findJsxElementAtNodePath = exports.extractStaticValue = exports.isStaticValue = void 0;
const node_fs_1 = require("node:fs");
const node_path_1 = __importDefault(require("node:path"));
const recast = __importStar(require("recast"));
const parse_ast_1 = require("../../codemods/parse-ast");
const isStaticValue = (node) => {
    switch (node.type) {
        case 'NumericLiteral':
        case 'StringLiteral':
        case 'BooleanLiteral':
        case 'NullLiteral':
            return true;
        case 'UnaryExpression':
            return ((node.operator === '-' ||
                node.operator === '+') &&
                node.argument.type === 'NumericLiteral');
        case 'ArrayExpression':
            return node.elements.every((el) => el !== null && el.type !== 'SpreadElement' && (0, exports.isStaticValue)(el));
        case 'ObjectExpression':
            return node.properties.every((prop) => prop.type === 'ObjectProperty' &&
                (0, exports.isStaticValue)(prop.value));
        default:
            return false;
    }
};
exports.isStaticValue = isStaticValue;
const extractStaticValue = (node) => {
    switch (node.type) {
        case 'NumericLiteral':
        case 'StringLiteral':
        case 'BooleanLiteral':
            return node.value;
        case 'NullLiteral':
            return null;
        case 'UnaryExpression': {
            const un = node;
            if (un.argument.type === 'NumericLiteral') {
                const val = un.argument.value;
                return un.operator === '-' ? -val : val;
            }
            return undefined;
        }
        case 'ArrayExpression':
            return node.elements.map((el) => {
                if (el === null || el.type === 'SpreadElement') {
                    return undefined;
                }
                return (0, exports.extractStaticValue)(el);
            });
        case 'ObjectExpression': {
            const obj = node;
            const result = {};
            for (const prop of obj.properties) {
                if (prop.type === 'ObjectProperty') {
                    const p = prop;
                    const key = p.key.type === 'Identifier'
                        ? p.key.name
                        : p.key.type === 'StringLiteral' ||
                            p.key.type === 'NumericLiteral'
                            ? String(p.key.value)
                            : undefined;
                    if (key !== undefined) {
                        result[key] = (0, exports.extractStaticValue)(p.value);
                    }
                }
            }
            return result;
        }
        default:
            return undefined;
    }
};
exports.extractStaticValue = extractStaticValue;
const getPropsStatus = (jsxElement) => {
    const props = {};
    for (const attr of jsxElement.attributes) {
        if (attr.type === 'JSXSpreadAttribute') {
            continue;
        }
        if (attr.name.type === 'JSXNamespacedName') {
            continue;
        }
        const { name } = attr.name;
        if (typeof name !== 'string') {
            continue;
        }
        const { value } = attr;
        if (!value) {
            props[name] = { canUpdate: true, codeValue: true };
            continue;
        }
        if (value.type === 'StringLiteral') {
            props[name] = {
                canUpdate: true,
                codeValue: value.value,
            };
            continue;
        }
        if (value.type === 'JSXExpressionContainer') {
            const { expression } = value;
            if (expression.type === 'JSXEmptyExpression' ||
                !(0, exports.isStaticValue)(expression)) {
                props[name] = { canUpdate: false, reason: 'computed' };
                continue;
            }
            props[name] = {
                canUpdate: true,
                codeValue: (0, exports.extractStaticValue)(expression),
            };
            continue;
        }
        props[name] = { canUpdate: false, reason: 'computed' };
    }
    return props;
};
const getNodePathForRecastPath = (
// eslint-disable-next-line @typescript-eslint/no-explicit-any
recastPath) => {
    const segments = [];
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    let current = recastPath;
    while (current && current.parentPath) {
        segments.unshift(current.name);
        current = current.parentPath;
    }
    // Recast paths start with "root" which doesn't correspond to a real AST property
    if (segments.length > 0 && segments[0] === 'root') {
        return segments.slice(1);
    }
    return segments;
};
const findJsxElementAtNodePath = (ast, nodePath) => {
    let current = new recast.types.NodePath(ast);
    for (const segment of nodePath) {
        current = current.get(segment);
        if (current.value === null || current.value === undefined) {
            return null;
        }
    }
    if (recast.types.namedTypes.JSXOpeningElement.check(current.value)) {
        return current.value;
    }
    return null;
};
exports.findJsxElementAtNodePath = findJsxElementAtNodePath;
const lineColumnToNodePath = (ast, targetLine) => {
    let foundPath = null;
    recast.types.visit(ast, {
        visitJSXOpeningElement(p) {
            const { node } = p;
            if (node.loc && node.loc.start.line === targetLine) {
                foundPath = getNodePathForRecastPath(p);
                return false;
            }
            return this.traverse(p);
        },
    });
    return foundPath;
};
exports.lineColumnToNodePath = lineColumnToNodePath;
const PIXEL_VALUE_REGEX = /^-?\d+(\.\d+)?px$/;
const isSupportedTranslateValue = (value) => {
    const parts = value.split(/\s+/);
    if (parts.length === 1 || parts.length === 2) {
        return parts.every((part) => PIXEL_VALUE_REGEX.test(part));
    }
    return false;
};
const validateStyleValue = (childKey, value) => {
    if (childKey === 'translate' && typeof value === 'string') {
        return isSupportedTranslateValue(value);
    }
    return true;
};
const getNestedPropStatus = (jsxElement, parentKey, childKey) => {
    const attr = jsxElement.attributes.find((a) => a.type !== 'JSXSpreadAttribute' &&
        a.name.type !== 'JSXNamespacedName' &&
        a.name.name === parentKey);
    if (!attr || !attr.value) {
        // Parent attribute doesn't exist, nested prop can be added
        return { canUpdate: true, codeValue: undefined };
    }
    if (attr.value.type !== 'JSXExpressionContainer') {
        return { canUpdate: false, reason: 'computed' };
    }
    const { expression } = attr.value;
    if (expression.type === 'JSXEmptyExpression' ||
        expression.type !== 'ObjectExpression') {
        // Parent is not an object literal (e.g. style={myStyles})
        return { canUpdate: false, reason: 'computed' };
    }
    const objExpr = expression;
    const prop = objExpr.properties.find((p) => p.type === 'ObjectProperty' &&
        ((p.key.type === 'Identifier' && p.key.name === childKey) ||
            (p.key.type === 'StringLiteral' && p.key.value === childKey)));
    if (!prop) {
        // Property not set in the object, can be added
        return { canUpdate: true, codeValue: undefined };
    }
    const propValue = prop.value;
    if (!(0, exports.isStaticValue)(propValue)) {
        return { canUpdate: false, reason: 'computed' };
    }
    const codeValue = (0, exports.extractStaticValue)(propValue);
    if (!validateStyleValue(childKey, codeValue)) {
        return { canUpdate: false, reason: 'computed' };
    }
    return { canUpdate: true, codeValue };
};
const computeSequencePropsStatusFromContent = (fileContents, nodePath, keys) => {
    const ast = (0, parse_ast_1.parseAst)(fileContents);
    const jsxElement = (0, exports.findJsxElementAtNodePath)(ast, nodePath);
    if (!jsxElement) {
        throw new Error('Could not find a JSX element at the specified location');
    }
    const allProps = getPropsStatus(jsxElement);
    const filteredProps = {};
    for (const key of keys) {
        const dotIndex = key.indexOf('.');
        if (dotIndex !== -1) {
            filteredProps[key] = getNestedPropStatus(jsxElement, key.slice(0, dotIndex), key.slice(dotIndex + 1));
        }
        else if (key in allProps) {
            filteredProps[key] = allProps[key];
        }
        else {
            filteredProps[key] = { canUpdate: true, codeValue: undefined };
        }
    }
    return {
        canUpdate: true,
        props: filteredProps,
        nodePath,
    };
};
exports.computeSequencePropsStatusFromContent = computeSequencePropsStatusFromContent;
const computeSequencePropsStatus = ({ fileName, nodePath, keys, remotionRoot, }) => {
    try {
        const absolutePath = node_path_1.default.resolve(remotionRoot, fileName);
        const fileRelativeToRoot = node_path_1.default.relative(remotionRoot, absolutePath);
        if (fileRelativeToRoot.startsWith('..')) {
            throw new Error('Cannot read a file outside the project');
        }
        const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
        return (0, exports.computeSequencePropsStatusFromContent)(fileContents, nodePath, keys);
    }
    catch (err) {
        return {
            canUpdate: false,
            reason: err.message,
        };
    }
};
exports.computeSequencePropsStatus = computeSequencePropsStatus;
const computeSequencePropsStatusByLine = ({ fileName, line, keys, remotionRoot, }) => {
    try {
        const absolutePath = node_path_1.default.resolve(remotionRoot, fileName);
        const fileRelativeToRoot = node_path_1.default.relative(remotionRoot, absolutePath);
        if (fileRelativeToRoot.startsWith('..')) {
            throw new Error('Cannot read a file outside the project');
        }
        const fileContents = (0, node_fs_1.readFileSync)(absolutePath, 'utf-8');
        const ast = (0, parse_ast_1.parseAst)(fileContents);
        const resolvedNodePath = (0, exports.lineColumnToNodePath)(ast, line);
        if (!resolvedNodePath) {
            throw new Error('Could not find a JSX element at the specified location');
        }
        return (0, exports.computeSequencePropsStatus)({
            fileName,
            nodePath: resolvedNodePath,
            keys,
            remotionRoot,
        });
    }
    catch (err) {
        return {
            canUpdate: false,
            reason: err.message,
        };
    }
};
exports.computeSequencePropsStatusByLine = computeSequencePropsStatusByLine;
