// @remotion/studio-server/dist/preview-server/routes/log-update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.logUpdate = void 0;
const renderer_1 = require("@remotion/renderer");
const make_link_1 = require("../../hyperlinks/make-link");
let warnedAboutPrettier = false;
const normalizeQuotes = (str) => {
    if (str.length >= 2 &&
        ((str.startsWith("'") && str.endsWith("'")) ||
            (str.startsWith('"') && str.endsWith('"')))) {
        return `'${str.slice(1, -1)}'`;
    }
    return str;
};
const formatValueChange = ({ oldValueString, newValueString, defaultValueString, }) => {
    // Changed to default value (prop gets deleted) → show only old value in red
    if (defaultValueString !== null && newValueString === defaultValueString) {
        return renderer_1.RenderInternals.chalk.red(oldValueString);
    }
    // Changed from default value (prop gets added) → show only new value in green
    if (defaultValueString !== null && oldValueString === defaultValueString) {
        return renderer_1.RenderInternals.chalk.green(newValueString);
    }
    return `${renderer_1.RenderInternals.chalk.red(oldValueString)} \u2192 ${renderer_1.RenderInternals.chalk.green(newValueString)}`;
};
const formatPropChange = ({ key, oldValueString, newValueString, defaultValueString, }) => {
    const isResetToDefault = defaultValueString !== null && newValueString === defaultValueString;
    const isChangeFromDefault = defaultValueString !== null && oldValueString === defaultValueString;
    const valueChange = formatValueChange({
        oldValueString,
        newValueString,
        defaultValueString,
    });
    const dotIndex = key.indexOf('.');
    if (dotIndex === -1) {
        if (isResetToDefault) {
            return renderer_1.RenderInternals.chalk.red(`${key}={${oldValueString}}`);
        }
        if (isChangeFromDefault) {
            return renderer_1.RenderInternals.chalk.green(`${key}={${newValueString}}`);
        }
        return `${key}={${valueChange}}`;
    }
    const parentKey = key.slice(0, dotIndex);
    const childKey = key.slice(dotIndex + 1);
    if (isResetToDefault) {
        return `${parentKey}={{${renderer_1.RenderInternals.chalk.red(`${childKey}: ${oldValueString}`)}}}`;
    }
    if (isChangeFromDefault) {
        return `${parentKey}={{${renderer_1.RenderInternals.chalk.green(`${childKey}: ${newValueString}`)}}}`;
    }
    return `${parentKey}={{${childKey}: ${valueChange}}}`;
};
const logUpdate = ({ absolutePath, fileRelativeToRoot, key, oldValueString, newValueString, defaultValueString, formatted, logLevel, }) => {
    const locationLabel = `${fileRelativeToRoot}`;
    const fileLink = (0, make_link_1.makeHyperlink)({
        url: `file://${absolutePath}`,
        text: locationLabel,
        fallback: locationLabel,
    });
    const propChange = formatPropChange({
        key,
        oldValueString: normalizeQuotes(oldValueString),
        newValueString: normalizeQuotes(newValueString),
        defaultValueString: defaultValueString !== null ? normalizeQuotes(defaultValueString) : null,
    });
    renderer_1.RenderInternals.Log.info({ indent: false, logLevel }, `${renderer_1.RenderInternals.chalk.blueBright(`${fileLink}:`)} ${propChange}`);
    if (!formatted && !warnedAboutPrettier) {
        warnedAboutPrettier = true;
        renderer_1.RenderInternals.Log.warn({ indent: false, logLevel }, renderer_1.RenderInternals.chalk.yellow('Could not format with Prettier. File will need to be formatted manually.'));
    }
};
exports.logUpdate = logUpdate;
