// @remotion/studio-server/dist/codemods/update-sequence-props.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateSequenceProps = void 0;
const recast = __importStar(require("recast"));
const can_update_sequence_props_1 = require("../preview-server/routes/can-update-sequence-props");
const parse_ast_1 = require("./parse-ast");
const update_nested_prop_1 = require("./update-nested-prop");
const b = recast.types.builders;
const updateSequenceProps = async ({ input, nodePath, key, value, defaultValue, prettierConfigOverride, }) => {
    var _a, _b;
    const ast = (0, parse_ast_1.parseAst)(input);
    let oldValueString = '';
    const isDefault = defaultValue !== null &&
        JSON.stringify(value) === JSON.stringify(defaultValue);
    const dotIndex = key.indexOf('.');
    const isNested = dotIndex !== -1;
    const parentKey = isNested ? key.slice(0, dotIndex) : key;
    const childKey = isNested ? key.slice(dotIndex + 1) : '';
    const node = (0, can_update_sequence_props_1.findJsxElementAtNodePath)(ast, nodePath);
    if (!node) {
        throw new Error('Could not find a JSX element at the specified line to update');
    }
    if (isNested) {
        oldValueString = (0, update_nested_prop_1.updateNestedProp)({
            node,
            parentKey,
            childKey,
            value,
            defaultValue,
            isDefault,
        });
    }
    else {
        const attrIndex = (_a = node.attributes) === null || _a === void 0 ? void 0 : _a.findIndex((a) => {
            if (a.type === 'JSXSpreadAttribute') {
                return false;
            }
            if (a.name.type === 'JSXNamespacedName') {
                return false;
            }
            return a.name.name === key;
        });
        const attr = attrIndex !== undefined && attrIndex !== -1
            ? (_b = node.attributes) === null || _b === void 0 ? void 0 : _b[attrIndex]
            : undefined;
        if (attr && attr.type !== 'JSXSpreadAttribute' && attr.value) {
            const printed = recast.print(attr.value).code;
            // Strip JSX expression container braces, e.g. "{30}" -> "30"
            oldValueString =
                printed.startsWith('{') && printed.endsWith('}')
                    ? printed.slice(1, -1)
                    : printed;
        }
        else if (attr && attr.type !== 'JSXSpreadAttribute' && !attr.value) {
            // JSX shorthand like `loop` (no value) is implicitly `true`
            oldValueString = 'true';
        }
        else if (!attr && defaultValue !== null) {
            oldValueString = JSON.stringify(defaultValue);
        }
        if (isDefault) {
            if (attr && attr.type !== 'JSXSpreadAttribute' && node.attributes) {
                node.attributes.splice(attrIndex, 1);
            }
        }
        else {
            const parsed = (0, update_nested_prop_1.parseValueExpression)(value);
            const newValue = value === true ? null : b.jsxExpressionContainer(parsed);
            if (!attr || attr.type === 'JSXSpreadAttribute') {
                const newAttr = b.jsxAttribute(b.jsxIdentifier(key), newValue);
                if (!node.attributes) {
                    node.attributes = [];
                }
                node.attributes.push(newAttr);
            }
            else {
                attr.value = newValue;
            }
        }
    }
    const finalFile = (0, parse_ast_1.serializeAst)(ast);
    let prettier = null;
    try {
        prettier = await Promise.resolve().then(() => __importStar(require('prettier')));
    }
    catch (_c) {
        return {
            output: finalFile,
            oldValueString,
            formatted: false,
        };
    }
    const { format, resolveConfig, resolveConfigFile } = prettier;
    let prettierConfig;
    if (prettierConfigOverride !== undefined) {
        prettierConfig = prettierConfigOverride;
    }
    else {
        const configFilePath = await resolveConfigFile();
        if (!configFilePath) {
            return {
                output: finalFile,
                oldValueString,
                formatted: false,
            };
        }
        prettierConfig = await resolveConfig(configFilePath);
    }
    if (!prettierConfig) {
        return {
            output: finalFile,
            oldValueString,
            formatted: false,
        };
    }
    const prettified = await format(finalFile, {
        ...prettierConfig,
        filepath: 'test.tsx',
        plugins: [],
        endOfLine: 'lf',
    });
    return {
        output: prettified,
        oldValueString,
        formatted: true,
    };
};
exports.updateSequenceProps = updateSequenceProps;
